src/main/java/de/tum/in/www1/artemis/service/ProgrammingExerciseService.java

```diff
@@ -324,55 +326,151 @@ public ProgrammingExercise setupProgrammingExercise(ProgrammingExercise programm
+            pushRepository(repository, templateName);
+        }
+    }
+
+    /**
+     * Set up the test repository. This method differentiates non sequential and sequential test repositories (more than 1 test job).
+     *
+     * @param repository
+     * @param resources
+     * @param prefix
+     * @param templateName
+     * @param programmingExercise
+     * @throws Exception
+     */
+    private void setupTestTemplateAndPush(Repository repository, Resource[] resources, String prefix, String templateName, ProgrammingExercise programmingExercise)
+            throws Exception {
+        if (gitService.listFiles(repository).size() == 0 && programmingExercise.getProgrammingLanguage() == ProgrammingLanguage.JAVA) { // Only copy template if repo is empty
+            String templatePath = "classpath:templates/" + programmingExercise.getProgrammingLanguage().toString().toLowerCase() + "/test";
+
+            String projectTemplatePath = templatePath + "/projectTemplate/**/*.*";
+            String testUtilsPath = templatePath + "/testutils/**/*.*";
+
+            Resource[] testUtils = ResourcePatternUtils.getResourcePatternResolver(resourceLoader).getResources(testUtilsPath);
+            Resource[] projectTemplate = ResourcePatternUtils.getResourcePatternResolver(resourceLoader).getResources(projectTemplatePath);
+
+            Map<String, Boolean> sectionsMap = new HashMap<>();
+
+            fileService.copyResources(projectTemplate, prefix, repository.getLocalPath().toAbsolutePath().toString(), false);
+
+            if (!programmingExercise.getSequentialTestRuns()) {
+                String testFilePath = templatePath + "/testFiles" + "/**/*.*";
+                Resource[] testFileResources = ResourcePatternUtils.getResourcePatternResolver(resourceLoader).getResources(testFilePath);
+
+                sectionsMap.put("non-sequential", true);
+                sectionsMap.put("sequential", false);
+
+                fileService.replacePlaceholderSections(Paths.get(repository.getLocalPath().toAbsolutePath().toString(), "pom.xml").toAbsolutePath().toString(), sectionsMap);
+
+                String packagePath = Paths.get(repository.getLocalPath().toAbsolutePath().toString(), "test", "${packageNameFolder}").toAbsolutePath().toString();
+                fileService.copyResources(testUtils, prefix, packagePath, true);
+                fileService.copyResources(testFileResources, prefix, packagePath, false);
             }
-            // there is no need in python to replace package names
+            else {
+                String stagePomXmlPath = templatePath + "/stagePom.xml";
+                Resource stagePomXml = ResourcePatternUtils.getResourcePatternResolver(resourceLoader).getResource(stagePomXmlPath);
+                // This is done to prepare for a feature where instructors/tas can add multiple build stages.
+                List<String> buildStages = new ArrayList<>();
```

Comment:
I don't like the variable name `buildStages`. What does it mean?

Reply:
I think the naming was not a good fit, it represents the test tasks in bamboo.
So we have e.g. one task for each `structural` and `behavior`.

I rename it to `sequentialTestTasks`. We can further discuss if you want.